flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/MessageAcknowledgingSourceBase.java

```diff
@@ -148,16 +149,12 @@ public void initializeState(FunctionInitializationContext context) throws Except
 		if (context.isRestored()) {
 			LOG.info("Restoring state for the {}.", getClass().getSimpleName());
 
-			List<SerializedCheckpointData[]> retrievedStates = new ArrayList<>();
+			List<Map<Long, Set<UId>>> retrievedStates = new ArrayList<>();
 			for (SerializedCheckpointData[] entry : this.checkpointedState.get()) {
-				retrievedStates.add(entry);
+				retrievedStates.add(SerializedCheckpointData.toDeque(entry, idSerializer));
 			}
+			pendingCheckpoints = SerializedCheckpointData.combine(retrievedStates);
 
-			// given that the parallelism of the function is 1, we can only have at most 1 state
```

Comment:
@aljoscha @kl0u Can you guys take a look at these changes here?
I'm pretty sure there's a fundamental reason why this source only supports a parallelism of 1.
I don't want to spend the time understanding exactly how this source base works, if you guys can briefly clarify if this change makes sense or not.

Reply:
To give some more context for these changes:

The MessageAcknowledingSourceBase can deduplicate messages it has already seen because it keeps a list of messageIds. The PubSubSource actually uses this too: https://github.com/apache/flink/pull/6594/files#diff-5d1317f139d297f4e67ff13defa4d66cR119

The thing is, because the list isn't shared and pubsub (nor RabbitMQ afaik) won't guarantee the message will be send to the same taskmanager this deduplication will only work with a parallelism of one.

In the case of this pubsub connector we only guarantee `at least once` so we like to be able to use a higher parallelism. It still tries to deduplicate but it is more of a best effort.

Though this actually does mean that when you're using the pubsub connector with a parallelism of 1 is becomes exactly-once.

Reply:
But doesn't it make more sense to disable the deduplication if the parallelism > 1 ? (We would not have the memory and CPU usage + smaller checkpoints)

If the user wants a best-effort deduping, they can implement it themselves in user code.

Reply:
Yep makes sense, I've made it conditional based on parallelism == 1